lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -533,55 +533,75 @@ swift::matchWitness(
   // SWIFT_ENABLE_TENSORFLOW
   auto result = finalize(anyRenaming, optionalAdjustments);
   if (result.isViable()) {
-    // '@differentiable' attributes must match completely. If there exists a
-    // '@differentiable' attribute with a superset of the "wrt" parameters of
-    // a requirement, then an '@differentiable' attribute is added
-    // automatically.
+    // For all `@differentiable` attributes of the protocol requirement, check
+    // that the witness has a derivative configuration with exactly the same
+    // parameter indices, or one with "superset" parameter indices. If there
+    // exists a witness derivative configuration with "superset" parameter
+    // indices, create an implicit `@differentiable` attribute for the witness
+    // with the exact parameter indices from the requirement `@differentiable`
+    // attribute.
     ASTContext &ctx = witness->getASTContext();
-    auto witnessDiffAttrs = witnessAttrs
-        .getAttributes<DifferentiableAttr, /*AllowInvalid*/ true>();
+    auto *witnessAFD = dyn_cast<AbstractFunctionDecl>(witness);
+    if (auto *witnessASD = dyn_cast<AbstractStorageDecl>(witness))
+      witnessAFD = witnessASD->getAccessor(AccessorKind::Get);
+    // NOTE: Validate `@differentiable` attributes by calling
+    // `getParameterIndices`. This is important for type-checking
+    // `@differentiable` attributes in non-primary files to skip invalid
+    // attributes and to resolve derivative configurations, used below.
+    for (auto *witnessDiffAttr :
+         witnessAttrs.getAttributes<DifferentiableAttr>()) {
+      (void)witnessDiffAttr->getParameterIndices();
+    }
     for (auto *reqDiffAttr : reqAttrs.getAttributes<DifferentiableAttr>()) {
-      // TODO(TF-482): Also check whether generic requirements are the same.
-      bool reqDiffAttrMatch = llvm::any_of(
-          witnessDiffAttrs, [&](const DifferentiableAttr *witnessDiffAttr) {
-            return witnessDiffAttr->getParameterIndices() &&
-                   reqDiffAttr->getParameterIndices() &&
-                   witnessDiffAttr->parametersMatch(*reqDiffAttr);
-          });
-      bool reqDiffAttrSupersetMatch = llvm::any_of(
-          witnessDiffAttrs, [&](const DifferentiableAttr *witnessDiffAttr) {
-            return witnessDiffAttr->getParameterIndices() &&
-                   reqDiffAttr->getParameterIndices() &&
-                   witnessDiffAttr->getParameterIndices()
-                     ->isSupersetOf(reqDiffAttr->getParameterIndices());
-          });
-      if (!reqDiffAttrMatch) {
-        auto implicitDiffAttr = false;
-        if (reqDiffAttrSupersetMatch) {
-          auto *witnessAFD = cast<AbstractFunctionDecl>(witness);
+      (void)reqDiffAttr->getParameterIndices();
```

Comment:
Note: I found that calling `DifferentiableAttr::getParameterIndices` on witness attributes was sufficient to make `test/AutoDiff/differentiable_attr_type_checking_primary_file.swift` pass. For good measure, I called it on requirement attributes too.